# `Linear Algebra using Numpy`

# <font color=red>Mr Fugu Data Science</font>

# (◕‿◕✿)

# `Purpose & Outcome:`

+ Learn new numpy operations
+ Practice linear algebra
    + brush up on how operations work by example

**Help Support the Channel: Buy Me A Coffee @mrfugudatasci**

[great resource](https://www.oreilly.com/library/view/machine-learning-with/9781491989371/ch01.html)

In [84]:
import numpy as np
np.__version__ # if you were wondering what version

'1.18.1'

# `Eigen values and vectors:`

<font size=4>**$Ax=\lambda x$**</font>, where **$\lambda$** is our eigen value of A
+ It deals with shrinking and stretching of vector ( x )
    + If the eigenvalue is zero, then we have a nullspace
+ the determinant <font size=4>$(A-\lambda I)=0$</font>


`------------------------------`


# Ex.)  $$A=\begin{bmatrix} 1 & 0 \\ -3 & 2  \end{bmatrix}$$

Let us solve this:

<font size=5>$det\begin{bmatrix} 1-\lambda & 0 \\ -3& 2-\lambda   \end{bmatrix}$</font>

<font size=5>$=(1-\lambda)(2-\lambda)-(3)*(0)$</font>

`factor this and use completing the squares`

<font size=5>$=\lambda^2-3\lambda+2$</font>

<font size=5>$=(\lambda-1)(\lambda-2)$</font>

<font size=5>$\lambda=1,2$</font> `finally our eigen values yippie.`

`---------------------------`

`Eigen Vectors`


solve for each w.r.t. $\lambda$

<font size=5>$\begin{bmatrix} 1-\lambda=1 & 0 \\ -3& 2-\lambda=1   \end{bmatrix}$</font> = <font size=5>$\begin{bmatrix} 0 & 0 \\ -3& 1   \end{bmatrix}$</font>

Now, you can solve with `Gaussian Elimination:`

<font size=5>$\begin{bmatrix} 0(3) & 0(3) \\ -3& 1   \end{bmatrix}$</font><font size=4>$\begin{bmatrix}v_{11}\\v_{1,2} \end{bmatrix}=0$</font>

<font size=4>$-3v_{1,1}+v_{1,2}=0$</font> 

<font size=4>$x_{11}=\frac{1}{3}v_{1,2}$</font>

<font size=4>$eigenVector(\lambda=1)=\begin{bmatrix}x_{1,1}\\\frac{1}{3}x_{1,2} \end{bmatrix}$</font>

`----------------------------`

`Now samething for` $\lambda=2$

<font size=5>$\begin{bmatrix} 1-\lambda=2 & 0 \\ -3& 2-\lambda=2   \end{bmatrix}$</font> = <font size=5>$\begin{bmatrix} -1 & 0 \\ -3& 0   \end{bmatrix}$</font>

Now, you can solve with `Gaussian Elimination:`

<font size=5>$\begin{bmatrix} -1(3) & 0(3) \\ -3& 0   \end{bmatrix}$</font><font size=4>$\begin{bmatrix}v_{11}\\v_{1,2} \end{bmatrix}=0$</font>

<font size=4>$-3v_{1,1}=0$</font> 

<font size=4>$x_{11}=0$</font>

<font size=4>$eigenVector(\lambda=2)=\begin{bmatrix}x_{1,1}=0\\x_{1,2}=? \end{bmatrix}$</font>

$x_{1,2}=?$, meaning that you can place a value since it will only be a multiple say (1) like the results.

[theory eigenvalue/vectors M.I.T help](http://math.mit.edu/~gs/linearalgebra/ila0601.pdf) | [matrix calculator with steps shown](https://matrixcalc.org/en/vectors.html#eigenvectors%28%7B%7B1,0%7D,%7B-3,2%7D%7D%29)

In [40]:
a = np.array([[1,0], [-3,2]])
eigenvalues, eigenvectors = np.linalg.eig(a)

eigenvalues,eigenvectors


(array([2., 1.]),
 array([[0.        , 0.31622777],
        [1.        , 0.9486833 ]]))

# `np.allclose( )` returns TRUE, if two arrays are element wise equal within a threshold 

In [39]:
# Verify e-vec * e-val = A * e-vec
d = eigenvectors[:,0] * eigenvalues[0]
e = a @ eigenvectors[:, 0]
np.allclose(d,e)

True

# `Hermitian Matrix:`

<font size=4>$A^{\theta}=A^T$</font>, the conjugate of (A) equals (A) transpose

$A=\begin{bmatrix} 6&2-j&4 \\ 2+j&-3&-j\\4&j&9   \end{bmatrix}$

`then`
$A^{\theta}=\begin{bmatrix} 6&2+j&4 \\ 2-j&-3&j\\4&-j&9   \end{bmatrix}$ = $A^T=\begin{bmatrix} 6&2+j&4 \\ 2-j&-3&j\\ 4&-j&9  \end{bmatrix}$

**Numpy Function:** `np.linalg.eigh( )`

In [8]:
a = np.array([[1,0], [-3,2]])
eigenvalues, eigenvectors = np.linalg.eigh(a)
eigenvalues, eigenvectors
np.linalg.eigh(a_)

(array([-1.54138127,  4.54138127]),
 array([[-0.76301998, -0.6463749 ],
        [-0.6463749 ,  0.76301998]]))

`------------------------------`

# `Trace: np.trace(  ) `
+ **Sum of the diagonals for a square matrix**

$tr(AB) = \begin{bmatrix}&\leftarrow \vec{a}_1\rightarrow& \\&\leftarrow \vec{a}_2\rightarrow&\\ &\vdots&  \\&\leftarrow\vec{a}_n\rightarrow& \end{bmatrix}$ $\begin{bmatrix}{\uparrow\\ {\vec{b}_1}\\ \downarrow }& {\uparrow\\{\vec{b}_2}\\\downarrow} & \dots & {\uparrow\\{\vec{b}_n}\\\downarrow}  \end{bmatrix}$
<font size=4> $=\begin{bmatrix}\vec{a}_1^T\vec{b}_1&\vec{a}_1^T\vec{b}_2&\dots&\vec{a}_1^T\vec{b}_n \\ \vec{a}_2^T\vec{b}_1&\vec{a}_2^T\vec{b}_2&\dots&\vec{a}_2^T\vec{b}_n \\   
\vdots&&\ddots&\vdots \\
\vec{a}_n^T\vec{b}_1&\vec{a}_n^T\vec{b}_2&\dots&\vec{a}_n^T\vec{b}_n
\end{bmatrix}$</font> = 
<font size=4>$\begin{bmatrix} \vec{a}_1^T\vec{b}_1&& \\
&\vec{a}_2^T\vec{b}_2& \\ &&\vec{a}_n^T\vec{b}_n
\end{bmatrix}$</font> = 
<font size=4>$\sum\limits_{i=1}^m\vec{a}_1^T\vec{b}_1 + \sum\limits_{i=1}^m
\vec{a}_2^T\vec{b}_2 +\sum\limits_{i=1}^m \vec{a}_n^T\vec{b}_n$</font>

[other properties of traces](https://web.stanford.edu/~jduchi/projects/matrix_prop.pdf)

In [45]:
import pandas as pd

df = pd.DataFrame({'a':[1,2,3,4],'b':[1,3,5,7],'c':[1,4,7,10],'d':[1,5,9,11]})

def highlight_diag(data, color='yellow'):
    '''
    highlight the diag values in a DataFrame
    '''
    attr = 'background-color: {}'.format(color)
    # create a new dataframe of the same structure with default style value
    df_style = data.replace(data, '')
    # fill diagonal with highlight color
    np.fill_diagonal(df_style.values, attr)
    return df_style

df.style.apply(highlight_diag, axis=None)

# code source in citations stackoverflow

In [59]:
# Trace of (A)
a_tr =np.array([[1,1,1,1],[2,3,4,5],[3,5,7,9],[4,7,10,11]])
np.trace(a_tr)

22

In [61]:
# alternate way to do same things: original trace
print(a_tr.trace())

print(sum(a_tr.diagonal()))

22
22


# `Offset trace:`

This allows you to work on the off-diagonals of your choice

In [54]:
# offset trace:
print('directly Right of yellow:',np.trace(a_tr,offset=1))
print('directly Left of yellow',np.trace(a_tr,offset=-1)) 

directly Right of yellow: 14
directly Left of yellow 17


# `Using Least Squares: np.linalg.lstsq(A,b)`
+ Under the hood uses SVD

When would we use this?

+ Well, if we had a system of equations outweighing amount of unknowns

`Ex.)` consider if we had these 3 equations:

$y=x_o+3x_1+x_2=6$

$y=2x_o+x_1+x_2=4$

$y=.2x_o-4x_1+.2x_2=7.4$

`We can express this as matrix: Ax=b`

+ `x: solution`
+ `residuals: error`
+ `rank:`
+ `s: singular values`

[one example](https://numpy.org/doc/stable/reference/generated/numpy.linalg.lstsq.html) | [example 02](https://riptutorial.com/numpy/example/16034/find-the-least-squares-solution-to-a-linear-system-with-np-linalg-lstsq)

In [150]:
A_l=np.array([[1,3,1],[2,1,1],[.2,-4,.2]])
b_l=np.array([6,4,7.4])
x, residuals, rank, s = np.linalg.lstsq(A_l,b_l,rcond=None)

# `Solving Linear Systems with: np.linalg.solve()`
+ (A) must be a square and full-rank matrix: All of its 
rows must be be linearly independent. 
+ (A) should be invertible/non-singular (its determinant is not zero)

# Ex.) two equations and two unknowns

$x_1+x_2=30$

$6x_1+2.3x_2=44$


we will use `Ax=b` to solve this problem


In [151]:
# Ax=b -> [A^-1]*b, but dont directly instead with .solve()
A = np.array([[1,1],[6,2.3]])
b= np.array([30,44])

np.linalg.solve(A,b)

array([-6.75675676, 36.75675676])

# `we could use: np.linalg.inv(A).dot(B)`

To solve the problem from above, but there is a caveat and considerations.
+ When using numpy we have two options to solve this problem, one is faster than the other.
Let's take into account our other option and how it can inhibit our computations/time.

+ We are starting with `Ax=b ->`$x=A^{-1}*b$ but computing the inverse is time consuming for a few reasons as well as resources allocated.

`Direct version`
Let's understand how `np.linalg.solve()` works.
+ we are not calculating the inverse, instead we are using LAPACK routine.
    + Then LU decomposition is used to find the values using forward/back substitution.
 
`Slower version & (can be) Not as accurate`
Now, `np.linalg.inv(A).dot(B)`
+ you are using more floating point operations to solve the inverse
    + If (A) is an ill-conditioned matrix you will have inaccuracy
    + useless steps that are unneeded while computing
 
[.solve() vs .inv.dot()](https://stackoverflow.com/questions/31256252/why-does-numpy-linalg-solve-offer-more-precise-matrix-inversions-than-numpy-li) | [LAPACK background](http://www.netlib.org/lapack/)

In [152]:
np.linalg.inv(A).dot(b)

array([-6.75675676, 36.75675676])

# `Multiple Linear Regression:`

<font size=4>$X =\begin{bmatrix} 1&x_{1,1}&x_{1,2}&\dots&x_{1,m} \\ 1&x_{2,1}&x_{2,2}&\dots&x_{2,m}\\\vdots&\vdots&\vdots&\ddots&x_{2,m}\\1&x_{n,1}&x_{n,2}&\dots&x_{n,m} \end{bmatrix}$,
$Y=\begin{bmatrix}y_1\\y_2\\\vdots\\y_n \end{bmatrix}$</font>


<font size=4>$\hat y = x\beta = \beta_o +\beta_1x+\beta_2x+...\beta_nx_n$</font>

<font size=4>$\beta=(X^TX)^{-1}X^Ty$</font>

[doing by hand with code example](http://www2.lawrence.edu/fast/GREGGJ/Python/numpy/numpyLA.html)

# Multiple Linear Regression: matrice form:

`Xt = np.transpose(X)
XtX = np.dot(Xt,X)
Xty = np.dot(Xt,y)
beta = np.linalg.solve(XtX,Xty)`

# `Matrix Multiplication:`

<font size=4>$A=\begin{bmatrix} 1&2&3&4 \\ 5&6&7&8\\9&10&11&12   \end{bmatrix}_{ 3x4}$
,$B=\begin{bmatrix} 2&4&6 \\ 8&10&12\\14&16&18  \end{bmatrix}_{ 3x3}$</font>

we have a $(3x4)(3x3)=(?)$, depends transpose and order of matrices `3x4 or 4x3`

<font size=4>$[A^TB]_{3x4}$ versus $[BA^T]_{3x4}$</font>

`Order MATTERS~`

# `np.multiply(): element wise multiplication`

In [122]:
# 4x3*3x3 = 4x3x3
aa=np.array([[1,2,3,4],[5,6,7,8],[9,10,11,12]])
bb=np.array([[2,4,6],[8,10,12],[14,16,18]])
np.multiply(aa[:,np.newaxis].T,bb)

# np.multiply(bb,aa[:,np.newaxis].T) # samething

array([[[  2,  20,  54],
        [  8,  50, 108],
        [ 14,  80, 162]],

       [[  4,  24,  60],
        [ 16,  60, 120],
        [ 28,  96, 180]],

       [[  6,  28,  66],
        [ 24,  70, 132],
        [ 42, 112, 198]],

       [[  8,  32,  72],
        [ 32,  80, 144],
        [ 56, 128, 216]]])

# Pay Attention:

# `np.matmul( ) != as np.multiply( )`

here are what the first few entrys of `np.matmul(bb,aa)` looks like

<font size=4>$B*A =\begin{bmatrix} (2+20+54=76)&x_{1,2}&\dots&x_{1,m} \\ (81+50+108=166)&x_{2,1}&x_{2,2}&\dots&x_{2,m}\\(14+80+162=256)&\vdots&\vdots&\ddots&x_{n,m}\\ \end{bmatrix}$

In [124]:
# 4x3*3x3 = 4*3
print(np.matmul(aa.T,bb))
print('------------------')
# 3x3*3x4 = 3x4 
print(np.matmul(bb,aa))

[[168 198 228]
 [192 228 264]
 [216 258 300]
 [240 288 336]]
------------------
[[ 76  88 100 112]
 [166 196 226 256]
 [256 304 352 400]]


# `Dot Product: np.dot(a,b) or a.dot(b)`

`Returns a Scalar`

<font size=4>$A=\begin{bmatrix} 2&3\\4&5\\  \end{bmatrix}$</font>,<font size=4>$B=\begin{bmatrix} 7&8\\9&10  \end{bmatrix}$</font>

<font size=4>$dot(AB)=\begin{bmatrix} 2*7+3*9&2*8+3*10 \\4*7+5*9&4*8+3*10  \end{bmatrix}$</font> = <font size=4>$\begin{bmatrix} 41&46\\73&62 \end{bmatrix}$</font>

In [74]:
a_=np.array([[2,3],[4,5]])
b_=np.array([[7,8],[9,10]])

print(np.dot(a_,b_))
print('---------')
print(a_.dot(b_))

[[41 46]
 [73 82]]
---------
[[41 46]
 [73 82]]


# `Inner Product: np.inner(a,b)` 

`Returns a SCALAR`

<font size=4>$A=\begin{bmatrix} 2&3\\4&5\\  \end{bmatrix}$</font>,<font size=4>$B=\begin{bmatrix} 7&8\\9&10  \end{bmatrix}$</font>

<font size=4>$dot(AB)=\begin{bmatrix} 2*7+3*8&2*9+3*10 \\4*7+5*8&4*9+5*10  \end{bmatrix}$</font> = <font size=4>$\begin{bmatrix} 38&48\\68&86 \end{bmatrix}$</font>


In [82]:
np.inner(a_,b_)

array([[38, 48],
       [68, 86]])

# `Cross Product:  np.cross( ) or a.cross(b)`

`Returns a VECTOR`

+ There is an assumption in numpy for horizontal vectors, so you may need to do a transpose if you get an error.

<font size=5>$$u = vxw = (x_1,x_2,x_3)(y_1,y_2,y_3)$$</font>

<font size=5>$$=(x_2y_3-x_3y_2)i+(x_3y_1-x_1y_3)j+(x_1y_2-x_2y_1)k$$</font>


In [83]:
np.cross(a_,b_)

array([-5, -5])

# `Outer Product: np.outer(a,b)`

`multiples (a) and (b) together`

<font size=4>$u \otimes v = uv^T =\begin{bmatrix}u_1\\u_2\\\vdots\\u_n \end{bmatrix} \begin{bmatrix}v_1&v_2&\dots&v_n  \end{bmatrix}=$</font><font size=4>$\begin{bmatrix} u_1v_1&u_1v_2&\dots&u_1v_n \\ u_2v_1&u_2v_2&\dots&u_2v_n \\ 
\vdots&\vdots&\ddots&\vdots \\ u_nv_1&u_nv_2&\dots&u_nv_n
\end{bmatrix}$</font>

[extra help for you](http://www.inf.ed.ac.uk/teaching/courses/cfcs1/lectures/cfcs_l10.pdf)

In [85]:
np.outer(a_,b_)

array([[14, 16, 18, 20],
       [21, 24, 27, 30],
       [28, 32, 36, 40],
       [35, 40, 45, 50]])

# <font color=red>Like</font>,Share &

# <font color=red>SUB</font>scribe

**`Help Support the channel: Buy Me A Coffee @mrfugudatasci`**

# `Citations & Help:`

# ◔̯◔

https://www.bogotobogo.com/python/python_numpy_matrix_tutorial.php

https://www.twilio.com/blog/2018/06/data-science-linear-algebra-python-scipy-numpy.html

http://www2.lawrence.edu/fast/GREGGJ/Python/numpy/numpyLA.html

https://web.stanford.edu/class/cs231a/section/section1.pdf

http://snowball.millersville.edu/~adecaria/ESCI386P/esci386-lesson18-Linear-Algebra.pdf

https://courses.cs.washington.edu/courses/cse446/20wi/Section1/linear_algebra.html

https://sites.calvin.edu/scofield/courses/m256/materials/eigenstuff.pdf

https://stackoverflow.com/questions/56916128/pandas-style-how-to-highlight-diagonal-elements

https://www.ms.uky.edu/~lee/amspekulin/eigenvectors.pdf